Dockerfile

```diff
@@ -1,16 +1,33 @@
-FROM ubuntu:22.04
+FROM ubuntu:22.04 AS builder
 
-WORKDIR /metachain 
+RUN apt update && apt upgrade -y
+RUN apt install -y curl
 
-# Requires copy the binary to `build` folder beforehand
-COPY build/* /metachain 
+WORKDIR /metachain
+RUN curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain none --profile minimal -y
+
+ENV PATH=$PATH:/root/.cargo/bin
+RUN apt install -y cmake clang
+
+COPY rust-toolchain.toml Cargo.toml Cargo.lock ./
+# TODO: docker layer cache, blocked by https://github.com/rust-lang/cargo/issues/2644
+# RUN cargo fetch
```

Comment:
Docker caching is currently not supported due to limitations of cargo/issues/2644. However, there are ways around this.

Ideally, the case scenario is to run cargo fetch on just the `Cargo.lock` file. However, that cannot run unless `src/*` are present which defeats the whole point of layered docker caching.

Alternatively:
- we could setup empty projects with mocked `src/*` structure and use `Cargo.lock` to `RUN cargo fetch`
- other use something like `cargo-chef`